prov/verbs_ud/src/verbs_ud.h

```diff
@@ -0,0 +1,738 @@
+	.protocol_version = 0					\
+}
+
+#define VERBS_UD_INJECT_FLAGS(ep, len, flags)			\
+	(((flags & FI_INJECT) ||				\
+	  (len <= ep->info->tx_attr->inject_size)) ?		\
+	 IBV_SEND_INLINE : 0)
+#define VERBS_UD_INJECT(ep, len)				\
+	VERBS__UD_INJECT_FLAGS(ep, len,				\
+			       ep->info->tx_attr->op_flags)
+
+#define VERBS_UD_SELECTIVE_COMP(ep)				\
+	(ep->ep_flags & FI_SELECTIVE_COMPLETION)
+#define VERBS_UD_COMP_FLAGS(ep, flags)				\
+	((!VERBS_UD_SELECTIVE_COMP(ep) ||			\
+	  (flags & (FI_COMPLETION | FI_TRANSMIT_COMPLETE))) ?	\
+	 IBV_SEND_SIGNALED : 0)
+#define VERBS_UD_COMP(ep)					\
+	VERBS_UD_COMP_FLAGS(ep, ep->info->tx_attr->op_flags)
+
+#define OFI_IBV_MTU_TYPE(bytes, ret_bytes)		\
+	case IBV_MTU_##bytes:			\
+		ret_bytes = bytes;		\
+		break
+
+static inline
+size_t ofi_verbs_ud_mtu_type_to_len(enum ibv_mtu mtu_type)
+{
+	size_t bytes = 0;
+
+	switch (mtu_type) {
+	OFI_IBV_MTU_TYPE(256, bytes);
+	OFI_IBV_MTU_TYPE(512, bytes);
+	OFI_IBV_MTU_TYPE(1024, bytes);
+	OFI_IBV_MTU_TYPE(2048, bytes);
+	OFI_IBV_MTU_TYPE(4096, bytes);
+	}
+
+	return bytes;
+}
+#define OFI_VERBS_UD_MTU_TYPE_TO_LEN(mtu_type)	\
+do {						\
+	switch ((enum ibv_mtu) mtu_type) {	\
+	case IBV_MTU_256:			\
+	case IBV_MTU_512:			\
+	case IBV_MTU_1024:			\
+	case IBV_MTU_2048:			\
+	case IBV_MTU_4096:			\
+	default:				\
+		assert(0);			\
+	}					\
+} while (0)
+
+struct ofi_verbs_ud_buf_pool;
+
+typedef void(*ofi_verbs_ud_pool_entry_cancel_hndlr) (struct ofi_verbs_ud_buf_pool *);
+
+struct ofi_verbs_ud_buf_pool {
+	struct util_buf_pool	*pool;
+	struct dlist_entry	buf_list;
+
+	ofi_verbs_ud_pool_entry_cancel_hndlr cancel_hndlr;
+};
+
+/*
+ * fields of Infiniband packet headers that are used to
+ * represent OFI EP address
+ * - LRH (Local Route Header) - Link Layer:
+ *   - LID - destination Local Identifier
+ *   - SL - Service Level
+ * - GRH (Global Route Header) - Network Layer:
+ *   - GID - destination Global Identifier
+ * - BTH (Base Transport Header) - Transport Layer:
+ *   - QPN - destination Queue Oair number
+ *   - PSN - Packet Serial Number
+ *   - P_key - Partition Key
+ *
+ * Note: DON'T change the placement of the fields in the structure.
+ *       The placement is to keep structure size = 256 bits (32 byte).
+ */
+struct ofi_ib_ud_ep_name {
+	union ibv_gid	gid; /* 64-bit GUID + 64-bit EUI - GRH */
+
+	uint32_t	qpn; /* BTH */
+	uint32_t	psn; /* SQ PSN - BTH */
+
+	uint16_t	lid; /* LRH */
+	uint16_t	pkey; /* BTH */
+	uint16_t	service; /* for NS src addr, 0 means any */
+
+	uint8_t 	sl; /* LRH */
+	uint8_t		padding; /* forced padding to 256 bits (32 byte) */
+}; /* 256 bits */
+
+struct ofi_verbs_ud_fabric {
+	struct util_fabric	util_fabric;
+	struct util_ns		name_server;
+};
+
+struct ofi_verbs_ud_domain {
+	struct util_domain	util_domain;
+	struct util_av		*av;
+    	struct fi_info		*info;
+	struct ibv_comp_channel	*ibv_comp_channel;
+	struct ibv_context	*ibv_device_ctx;
+	int			port_number;
+	struct ibv_port_attr	*port_attr;
```

Comment:
Yes, you're right - OFI domain maps to the <device, port> pair. Do you think that this is excessive?
Doing such mapping we generates fi_info for all ports of the device (not only for port with number 1 - as it is implemented in verbs/MSG,RDM).
I think it can be useful in the future. If some device has 2 active ports with 10 GB/s bandwidth per port, user may utilize 2 OFI domains and got overall bandwidth = 20 GB/s. What do you think, is it make sense?